File: src/chips/SN74LS161A.cpp
#include "chips/SN74LS161A.h"

#include <bitset>

using namespace IcSim;

SN74LS161A::SN74LS161A(std::string name) : ChipBase(name)
{
    // Init pins
    PinNameMap pins_ = {
        {"MR",  1},
        {"CP",  2},
        {"P0",  3},
        {"P1",  4},
        {"P2",  5},
        {"P3",  6},
        {"CEP", 7},
        {"GND", 8},
        {"PE",  9},
        {"CET", 10},
        {"Q3",  11},
        {"Q2",  12},
        {"Q1",  13},
        {"Q0",  14},
        {"TC",  15},
        {"VCC", 16},
    };
    set_pins(pins_);

    // Init data
    set_data("Q0", 0);
    set_data("Q1", 0);
    set_data("Q2", 0);
    set_data("Q3", 0);
    set_data("TC", 0);
}

void
SN74LS161A::cook(bool& prop)
{
    // The master reset will happen any time MR is low and is not tied to clock
    if (get_input(1) == 0) {
        count_ = 0;
        set_data(14, 0, prop);
        set_data(13, 0, prop);
        set_data(12, 0, prop);
        set_data(11, 0, prop);
        set_data(15, 0, prop);
        return;
    }
    int clk = get_input(2);
    if (clk && !clocked_) {
        // If PE is low, then load the data_ on the P inputs into the Q outputs and calculate the int count from them
        if (get_input(9) == 0) {
            const int b0 = get_input(3);
            set_data(14, b0, prop);
            const int b1 = get_input(4);
            set_data(13, b1, prop);
            const int b2 = get_input(5);
            set_data(12, b2, prop);
            const int b3 = get_input(6);
            set_data(11, b3, prop);
            set_data(15, 0, prop);
            std::bitset<4> count;
            count[0] = b0;
            count[1] = b1;
            count[2] = b2;
            count[3] = b3;
            count_ = count.to_ullong();
            clocked_ = true;
            return;
        }
        // Reset the carry flag after a full rise and fall clock
        if (output_value(15) == 1) {
            set_data(15, 0, prop);
        }
        // If CET and CEP are true we count
        const int cet = get_input(10);
        const int cep = get_input(7);
        if (cet && cep) {
            count_ += 1;
            clocked_ = true;

            // Roll over
            if (count_ == 16) {
                set_data(15, 1, prop);
                count_ = 0;
            }

            std::bitset<4> count(count_);
            set_data(14, count[0], prop);
            set_data(13, count[1], prop);
            set_data(12, count[2], prop);
            set_data(11, count[3], prop);
        }
    }

    if (clocked_ && !clk) {
        clocked_ = false;
    }
}

File: py/SN74LS161A.py
from chip_base import ChipBase
from exceptions import NotConnected

from bitstring import BitArray

class SN74LS161A(ChipBase):
    def __init__(self, name):
        super(SN74LS161A, self).__init__(name)

        self.type = 'SN74LS161A'
        self.discription = "4 bit binary counter with synchronous reset and parallel or serial input"
        self.has_clock_input = True

        self.pins = {
            'MR': 1,
            'CP': 2,
            'P0': 3,
            'P1': 4,
            'P2': 5,
            'P3': 6,
            'CEP': 7,
            'GND': 8,
            'PE': 9,
            'CET': 10,
            'Q3': 11,
            'Q2': 12,
            'Q1': 13,
            'Q0': 14,
            'TC': 15,
            'VCC': 16,
        }

        self.data = {
            'Q0': 0,
            'Q1': 0,
            'Q2': 0,
            'Q3': 0,
            'TC': 0,
        }

        self._count = 0

        self._clocked = False

    def cook(self):
        # The master reset will happen any time MR is low and is not tied to clock
        if self.get_input('MR') == 0:
            self._count = 0
            self.data['Q0'] = 0
            self.data['Q1'] = 0
            self.data['Q2'] = 0
            self.data['Q3'] = 0
            self.data['TC'] = 0
            return

        clk = self.get_input('CP')
        if clk and (not self._clocked):
            # If PE is low, then load the data on the P inputs into the Q outputs and calculate the int count from them
            if self.get_input('PE') == 0:
                self.data['Q0'] = self.get_input('P0')
                self.data['Q1'] = self.get_input('P1')
                self.data['Q2'] = self.get_input('P2')
                self.data['Q3'] = self.get_input('P3')
                self.data['TC'] = 0
                self._count = BitArray(bin="{}{}{}{}".format(self.data['Q3'], self.data['Q2'], self.data['Q1'], self.data['Q0'])).uint
                self._clocked = True
                return
            # Reset the carry flag after a full rise and fall clock
            if self.data.get('TC'):
                self.data['TC'] = 0
            # If CET and CEP are true we count
            cet = self.get_input('CET')
            cep = self.get_input('CEP')
            if cet and cep:
                self._count += 1
                self._clocked = True
        
        if self._clocked and (not clk):
            self._clocked = False

        # Roll over
        if self._count == 16:
            self._count = 0
            self.data['Q0'] = 0
            self.data['Q1'] = 0
            self.data['Q2'] = 0
            self.data['Q3'] = 0
            self.data['TC'] = 1
        
        # Get a lis of binary numbers from the count
        res = [int(i) for i in list('{0:04b}'.format(self._count))]
        self.data['Q0'] = res[3]
        self.data['Q1'] = res[2]
        self.data['Q2'] = res[1]
        self.data['Q3'] = res[0]

    def __str__(self):
        return "{} {} {} {}".format(self.data.get('Q3'),self.data.get('Q2'),self.data.get('Q1'),self.data.get('Q0'))
